Modules/_ctypes/cfield.c

```diff
@@ -31,40 +31,159 @@ static void pymem_destructor(PyObject *ptr)
+PyCField_FromDesc_msvc to simulate the corresponding compilers.
+
+GCC mode places fields one after another, bit by bit.  But when a field would
+straddle an alignment boundary for its type, we insert a few bits of padding to
+avoid that.
+
+MSVC mode works similar except for bitfield packing.  Adjacent bit-fields are
+packed into the same 1-, 2-, or 4-byte allocation unit if the integral types
+are the same size and if the next bit-field fits into the current allocation
+unit without crossing the boundary imposed by the common alignment requirements
+of the bit-fields.
+
+See https://gcc.gnu.org/onlinedocs/gcc/x86-Options.html#index-mms-bitfields for details.
+
+We do not support zero length bitfields.  In fact we use bitsize != 0 to
+indicate a bitfield.
+
+PyCField_FromDesc manages:
+- *psize: the size of the structure / union so far.
+- *poffset, *pbitofs: 8* (*poffset) + *pbitofs points to where the next field
+  would start.
+- *palign: the alignment requirements of the last field we placed.
+*/
+
+static void
+PyCField_FromDesc_gcc(Py_ssize_t bitsize, Py_ssize_t *pbitofs,
                 Py_ssize_t *psize, Py_ssize_t *poffset, Py_ssize_t *palign,
-                int pack, int big_endian)
+                CFieldObject* self, StgInfo* dict,
+                int is_bitfield
+                )
 {
-    CFieldObject *self;
-    PyObject *proto;
-    Py_ssize_t size, align;
-    SETFUNC setfunc = NULL;
-    GETFUNC getfunc = NULL;
-    int fieldtype;
-#define NO_BITFIELD 0
-#define NEW_BITFIELD 1
-#define CONT_BITFIELD 2
-#define EXPAND_BITFIELD 3
+    // We don't use poffset here, so clear it, if it has been set.
+    *pbitofs += *poffset * 8;
+    *poffset = 0;
+
+    *palign = dict->align;
+
+    if ((bitsize > 0)
+        && (round_down(*pbitofs, 8 * dict->align)
+            < round_down(*pbitofs + bitsize - 1, 8 * dict->align))) {
+        // We would be straddling alignment units.
+        *pbitofs = round_up(*pbitofs, 8*dict->align);
+    }
+    assert(*poffset == 0);
+
+    if(is_bitfield) {
+        self->offset = round_down(*pbitofs, 8*dict->size) / 8;
+        Py_ssize_t effective_bitsof = *pbitofs - 8 * self->offset;
+        self->size = BUILD_SIZE(bitsize, effective_bitsof);
+        assert(dict->size == dict->align);
+        assert(effective_bitsof <= dict->size * 8);
+    } else {
+        self->offset = round_down(*pbitofs, 8*dict->align) / 8;
+        self->size = dict->size;
+    }
+
+    *pbitofs += bitsize;
+    *psize = round_up(*pbitofs, 8) / 8;
+}
+
+static void
+PyCField_FromDesc_msvc(
+                Py_ssize_t *pfield_size, Py_ssize_t bitsize,
+                Py_ssize_t *pbitofs, Py_ssize_t *psize, Py_ssize_t *poffset,
+                Py_ssize_t *palign, int pack,
+                CFieldObject* self, StgInfo* dict,
+                int is_bitfield
+                )
+{
+    if (pack)
+        *palign = min(pack, dict->align);
+    else
+        *palign = dict->align;
```

Comment:
* PEP-7 nitpick: use braces for all blocks.

* AFAIK, the `min` define is a ctypes-specific accident of history; please use CPython's `Py_MIN`.

```suggestion
    if (pack) {
        *palign = Py_MIN(pack, dict->align);
    } else {
        *palign = dict->align;
    }
```

(I see this is moved, rather than new code. Still, probably better to touch it up.)